extensions/guacamole-vault/modules/guacamole-vault-ksm/src/main/java/org/apache/guacamole/vault/ksm/user/KsmDirectoryService.java

```diff
@@ -0,0 +1,241 @@
+
+                    // Create an options object using the values we extracted from the one-time token
+                    SecretsManagerOptions options = new SecretsManagerOptions(
+                        storage, null,
+                        configurationService.getAllowUnverifiedCertificate());
+
+                    // Attempt to fetch secrets using the options we created. This will both validate
+                    // that the configuration works, and potentially populate missing fields that the
+                    // initializeStorage() call did not set.
+                    SecretsManager.getSecrets(options);
+
+                    // Create a map to store the extracted values from the KSM storage
+                    Map<String, String> configMap = new HashMap<>();
+
+                    // Go through all the expected fields, extract from the KSM storage,
+                    // and write to the newly created map
+                    EXPECTED_KSM_FIELDS.forEach(configKey -> {
+
+                        // Only write the value into the new map if non-null
+                        String value = storage.getString(configKey);
+                        if (value != null)
+                            configMap.put(configKey, value);
+
+                    });
+
+                    // JSON-encode the value, and then base64 encode that to get the format
+                    // that KSM would expect
+                    String jsonString = new ObjectMapper().writeValueAsString(configMap);
+                    String base64EncodedJson = Base64.getEncoder().encodeToString(jsonString.getBytes());
+
+                    // Finally, try to parse the newly generated token as a KSM config. If this
+                    // works, the config should be fully functional
+                    KsmConfig.parseKsmConfig(base64EncodedJson);
+
+                    // Make a copy of the existing attributes, modifying just the value for
+                    // KSM_CONFIGURATION_ATTRIBUTE
+                    attributes = new HashMap<>(entity.getAttributes());
+                    attributes.put(
+                            KsmAttributeService.KSM_CONFIGURATION_ATTRIBUTE, base64EncodedJson);
+
+                    // Set the newly updated attributes back to the original object
+                    entity.setAttributes(attributes);
+
+                }
+
+                // The KSM SDK only throws raw Exceptions, so we can't be more specific
+                catch (Exception exception) {
+
+                    // If the parsing attempt fails, throw a translatable error for display
+                    // on the frontend
+                    throw new TranslatableGuacamoleClientException(
+                            "Invalid one-time KSM token provided for "
+                            + KsmAttributeService.KSM_CONFIGURATION_ATTRIBUTE + " attribute",
+                            "CONNECTION_GROUP_ATTRIBUTES.ERROR_INVALID_KSM_ONE_TIME_TOKEN");
```

Comment:
Here, too - the underlying exception should be included in the constructor as the cause.